sled-agent/src/support_bundle.rs

```diff
@@ -0,0 +1,291 @@
+            }
+        })?;
+
+    Ok(SupportBundleCmdOutput { command: cmd_string, stdio, exit_status })
+}
+
+// This internal implementation method accepts a oneshot channel which is mostly
+// useful for unit tests so that we can ensure the blocking task has finished
+// appropriately when the spawned command has finished or has been killed.
+async fn execute_command_with_timeout_impl(
+    command: Command,
+    oneshot: oneshot::Sender<()>,
+    duration: Duration,
+) -> Result<SupportBundleCmdOutput, SupportBundleCmdError> {
+    let cmd_string = command_to_string(&command);
+    let tokio_command = execute(command, oneshot);
+    match tokio::time::timeout(duration, tokio_command).await {
+        Ok(res) => res,
+        Err(_elapsed) => Err(SupportBundleCmdError::Timeout {
+            command: cmd_string,
+            duration,
+        }),
+    }
+}
+
+/// Spawn a command that's allowed to execute within a given time limit.
+async fn execute_command_with_timeout(
+    command: Command,
+    duration: Duration,
+) -> Result<SupportBundleCmdOutput, SupportBundleCmdError> {
+    let (tx, _rx) = oneshot::channel();
+    execute_command_with_timeout_impl(command, tx, duration).await
+}
+
+fn zoneadm_list() -> Command {
+    let mut cmd = std::process::Command::new(PFEXEC);
+    cmd.env_clear().arg(ZONEADM).arg("list").arg("-cip");
+    cmd
+}
+
+fn ipadm_show_interface() -> Command {
+    let mut cmd = std::process::Command::new(PFEXEC);
+    cmd.env_clear().arg(IPADM).arg("show-if");
+    cmd
+}
+
+fn ipadm_show_addr() -> Command {
+    let mut cmd = std::process::Command::new(PFEXEC);
+    cmd.env_clear().arg(IPADM).arg("show-addr");
+    cmd
+}
+
+fn ipadm_show_prop() -> Command {
+    let mut cmd = std::process::Command::new(PFEXEC);
+    cmd.env_clear().arg(IPADM).arg("show-prop");
+    cmd
+}
+
+/*
+ * Public API
+ */
+
+/// List all zones on a sled.
+pub async fn zoneadm_info(
+) -> Result<SupportBundleCmdOutput, SupportBundleCmdError> {
+    execute_command_with_timeout(zoneadm_list(), DEFAULT_TIMEOUT).await
+}
+
+/// Retrieve various `ipadm` command output for the system.
+pub async fn ipadm_info(
+) -> Vec<Result<SupportBundleCmdOutput, SupportBundleCmdError>> {
+    [ipadm_show_interface(), ipadm_show_addr(), ipadm_show_prop()]
+        .into_iter()
+        .map(|c| async move {
+            execute_command_with_timeout(c, DEFAULT_TIMEOUT).await
+        })
+        .collect::<FuturesUnordered<_>>()
+        .collect::<Vec<Result<SupportBundleCmdOutput, SupportBundleCmdError>>>()
+        .await
+}
+
+#[cfg(test)]
+mod test {
+    use std::{process::Command, time::Duration};
+
+    use crate::support_bundle::*;
+
+    #[tokio::test]
+    async fn test_drop_cleans_up_spawn_blocking() {
+        let mut command = Command::new("sleep");
+        command.env_clear().arg("10");
+
+        let (tx, mut rx) = oneshot::channel();
+        let handle = tokio::spawn(async {
+            execute_command_with_timeout_impl(
```

Comment:
See my comments above - I think I understand why this test exists - you want to check that the `spawn_blocking` task doesn't exist after we bail out - but IMO it would be nice to avoid a "test-only" oneshot, if we can.

If I'm wrong, and we really do need the oneshot, we could consider wrapping it in `#[cfg(test)]` annotations to avoid unnecessary message passing / allocations.